alts/src/main/java/io/grpc/alts/AltsServerBuilder.java

```diff
@@ -182,77 +200,27 @@ public TsiHandshaker newHandshaker() {
                     new AltsHandshakerOptions(RpcProtocolVersionsUtil.getRpcProtocolVersions()));
               }
             }));
-    return new AltsServer(delegate.build());
+    return delegate.intercept(new AltsServerInterceptor(status)).build();
   }
 
-  static final class AltsServer extends io.grpc.Server {
-    private final Server delegate;
-
-    AltsServer(Server delegate) {
-      this.delegate = delegate;
-    }
-
-    @Override
-    public List<ServerServiceDefinition> getImmutableServices() {
-      return delegate.getImmutableServices();
-    }
-
-    @Override
-    public List<ServerServiceDefinition> getMutableServices() {
-      return delegate.getMutableServices();
-    }
-
-    @Override
-    public int getPort() {
-      return delegate.getPort();
-    }
-
-    @Override
-    public List<ServerServiceDefinition> getServices() {
-      return delegate.getServices();
-    }
-
-    @Override
-    public Server start() throws IOException {
-      delegate.start();
-      return this;
-    }
+  /** An implementation of {@link ServerInterceptor} that fails each call if status is not OK. */
+  static final class AltsServerInterceptor implements ServerInterceptor {
 
-    @Override
-    public Server shutdown() {
-      delegate.shutdown();
-      return this;
-    }
+    private final Status status;
 
-    @Override
-    public Server shutdownNow() {
-      delegate.shutdownNow();
-      return this;
-    }
-
-    @Override
-    public boolean isShutdown() {
-      return delegate.isShutdown();
-    }
-
-    @Override
-    public boolean isTerminated() {
-      return delegate.isTerminated();
-    }
-
-    @Override
-    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
-      return delegate.awaitTermination(timeout, unit);
-    }
-
-    @Override
-    public void awaitTermination() throws InterruptedException {
-      delegate.awaitTermination();
+    public AltsServerInterceptor(Status status) {
+      this.status = status;
     }
 
     @Override
-    public String toString() {
-      return MoreObjects.toStringHelper(this).add("delegate", delegate).toString();
+    public <ReqT, RespT> Listener<ReqT> interceptCall(
+        ServerCall<ReqT, RespT> serverCall,
+        Metadata metadata,
+        ServerCallHandler<ReqT, RespT> nextHandler) {
+      if (!status.isOk()) {
+        serverCall.close(status, new Metadata());
```

Comment:
This would be appropriate, although, isn't it impossible to actually get here? Won't the server be unable to accept new connections? And without connections it won't get any RPCs.

Server-side does have `start()` that can throw an `IOException`...